confgenerator/prometheus.go

```diff
@@ -84,15 +86,62 @@ func (r PrometheusMetrics) Pipelines() []otel.Pipeline {
 		}
 	}
 
-	// TODO(b/248268653): Fix the issue we have with the regex capture group syntax.
 	return []otel.Pipeline{{
-		Receiver: otel.Component{
-			Type:   "prometheus",
-			Config: map[string]interface{}{"config": r.PromConfig},
-		},
+		Receiver: prometheusToOtelComponent(r.PromConfig),
 	}}
 }
 
+// Generate otel components for the prometheus config used. It is the same config except
+// we need to escape the $ characters in the regexes.
+//
+// Note: We copy over the prometheus scrape configs and create new ones so calls to `Pipelines()`
+// will return the same result everytime and not change the original prometheus config.
+func prometheusToOtelComponent(promConfig promconfig.Config) otel.Component {
+	copyPromConfig, err := deepCopy(promConfig)
+	if err != nil {
+		// This should never happen since we already validated the prometheus config.
+		panic(fmt.Errorf("failed to deep copy prometheus config: %w", err))
+	}
+
+	for i := range promConfig.ScrapeConfigs {
+		for j := range copyPromConfig.ScrapeConfigs[i].RelabelConfigs {
+			// Escape the $ characters in the regexes.
+			replacement := copyPromConfig.ScrapeConfigs[i].RelabelConfigs[j].Replacement
+			replacement = strings.ReplaceAll(replacement, "$", "$$")
+			copyPromConfig.ScrapeConfigs[i].RelabelConfigs[j].Replacement = replacement
+		}
+		for j := range copyPromConfig.ScrapeConfigs[i].MetricRelabelConfigs {
+			// Escape the $ characters in the regexes.
+			replacement := copyPromConfig.ScrapeConfigs[i].MetricRelabelConfigs[j].Replacement
+			replacement = strings.ReplaceAll(replacement, "$", "$$")
+			copyPromConfig.ScrapeConfigs[i].MetricRelabelConfigs[j].Replacement = replacement
+		}
+	}
+
+	return otel.Component{
+		Type:   "prometheus",
+		Config: map[string]interface{}{"config": copyPromConfig},
+	}
+}
+
+func deepCopy(config promconfig.Config) (promconfig.Config, error) {
+	var out any
+	if err := mapstructure.Decode(config, &out); err != nil {
+		return promconfig.Config{}, err
+	}
```

Comment:
Question: why do we need to convert the config to any here?

Reply:
Good catch. We don't need the `Decode` at all it looks like, this was from when I copied over the yaml marshalling that we do in the `ModularConfig.Generate()` but we can make this simpler. Done in latest commit.